x-pack/plugins/fleet/server/services/setup.ts

```diff
@@ -68,15 +68,42 @@ export async function setupFleet(
   esClient: ElasticsearchClient
 ): Promise<SetupStatus> {
   const t = apm.startTransaction('fleet-setup', 'fleet');
+  const logger = appContextService.getLogger();
 
   try {
+    // check if fleet setup is already started
+    const settings = await settingsService.getSettingsOrUndefined(soClient);
+
+    if (settings && settings.fleet_setup_status === 'in_progress') {
```

Comment:
I think there is still a possibility for a race condition no? I am wondering if we should introduce a real distributed lock for the setup.

Not sure how easy it will be to implement this with saved object, maybe something like this with no retry on conflict, happy to discuss this more
```
await settingsService.saveSettings(soClient, {
        fleet_setup_status: 'in_progress',
    fleet_setup_id: uuid(), // to check who get the lock
    fleet_setup_started_at: Date.now(), // to get a TTL for the lock
      });
```

Reply:
+1, what are the exact concurrency semantics here? Is this actually atomic? A distributed lock would work if you had a perfect one, but those can also run into problems around retries depending on exactly how it works.

A better solution would be to make it so that concurrent attempts to create the policy don't matter by making the action idempotent. What if the requests to create the policy went through a queue as a way of serializing the action and getting a consistent result each time? 

What if each instance of Kibana writes to it's own copy of the same object with a unique name, and we pick the first one as the one that gets used, or an arbitrary one, assuming they are all the same.

There might be other options as well.

Reply:
My thinking with the in progress flag was to avoid the unnecessary work of each kibana instance creating the preconfigured policies. We are using a similar approach in package install where the package is set to `installing` status and can't be installed concurrently until completed.

Something like selecting a leader kibana instance could work, though I'm not sure how easy it is to implement.

Using a queue would need a separate component that executes it, isn't it? 

Using a unique policy id for each attempt would potentially generate a lot of documents, as Fleet setup runs every time Kibana restarts or Fleet UI is visited. We are also relying on a specific name of the cloud policy.

I think a less intrusive change would be to catch version conflict errors and check the existence of the policy before retrying/rollback. 
Also how we are bumping the revision field now is not atomic, it is being read and updated separately, we could improve it with an update painless script, to make sure two instances are not overwriting the same revision doc.

Reply:
@juliaElastic If we go to avoid concurrent calls could we catch the version conflict when updating the settings object, (and use that document as the lock)?  

> Using a queue would need a separate component that executes it, isn't it?

We could eventually use the task manager for that, that is the available queue in Kibana, but it will add some delay to bump the policy not sure if it's acceptable for us. But it could solve a few scalabilty issue to when bumping agent policies revision.